tools/crm_rule.c

```diff
@@ -0,0 +1,266 @@
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS, data_set->input);
+
+    /* Get all rules matching the given ID, which also have only a single date_expression
+     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
+     * to check expressions more complicated than that.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+
+    max = numXpathResults(xpathObj);
+    if (max == 0) {
+        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        rc = 0;
+        goto bail;
+    } else if (max > 1) {
+        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        rc = CRM_EX_ERROR;
+        goto bail;
+    }
+
+    match = getXpathResult(xpathObj, 0);
+
+    /* We should have ensured both of these pass with the xpath query above, but
+     * double checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+    CRM_ASSERT(find_expression_type(match) == time_expr);
+
+    result = pe_eval_date_expression(match, effective_date);
+
+    if (result == date_after_range) {
+        printf("Rule %s is expired\n", rule_id);
+        rc = 1;
+    } else if (result == date_within_range) {
+        printf("Rule %s is still in effect\n", rule_id);
+        rc = 0;
+    } else if (result == date_before_range) {
+        printf("Rule %s has not yet taken effect\n", rule_id);
+        rc = 1;
+    } else {
+        printf("Could not determine whether rule %s is expired\n", rule_id);
+        rc = 1;
+    }
+
+bail:
+    free(xpath);
+    freeXpathObject(xpathObj);
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+    cib_t *cib_conn = NULL;
+
+    int flag = 0;
+    int option_index = 0;
+
+    char *rule_id = NULL;
+    crm_time_t *rule_date = NULL;
+
+    xmlNode *input = NULL;
+    char *input_xml = NULL;
+    char *input_file = getenv("CIB_file");
+
+    int rc = pcmk_ok;
+    crm_exit_t exit_code = CRM_EX_OK;
+
+    crm_log_cli_init("crm_rule");
+    crm_set_options(NULL, "[options]", long_options,
+                    "Tool for querying the state of rules");
+
+    while (flag >= 0) {
+        flag = crm_get_option(argc, argv, &option_index);
+        switch (flag) {
+            case -1:
+                break;
+
+            case 'V':
+                crm_bump_log_level(argc, argv);
+                break;
+
+            case '$':
+            case '?':
+                crm_help(flag, CRM_EX_OK);
+                break;
+
+            case 'c':
+                rule_mode = crm_rule_mode_check;
+                break;
+
+            case 'd':
+                rule_date = crm_time_new(optarg);
+                if (rule_date == NULL) {
+                    exit_code = CRM_EX_DATAERR;
+                    goto bail;
```

Comment:
This is a correct example; exit_code is being set directly to one of the exit code constants.